incipyt/hooks.py

```diff
@@ -18,11 +22,75 @@ def register(cls, hook):
     def __init__(self, hierarchy):
         self._hierarchy = hierarchy
 
+    def __str__(self):
+        return "build-dependancy"
+
+    def __call__(self, value):  # noqa: D102
+        if self._hooks:
+            logger.info(f"{len(self._hooks)} {self} hooks called with '{value}'")
+
+        for hook in self._hooks:
+            hook(self._hierarchy, value)
+
+
+class Classifier:
+    """Hook to add a classifier to a build system."""
+
+    _hooks = []
+
+    @classmethod
+    def register(cls, hook):
+        """Regsiter a callback for :class:`Classifier` hook.
+
+        :param hook: Callaback to register.
+        :type hook: :class:`collections.abc.Callable`
+        """
+        cls._hooks.append(hook)
+
+    def __init__(self, hierarchy):
+        self._hierarchy = hierarchy
+
+    def __str__(self):
+        return "classifier"
+
     def __call__(self, value):  # noqa: D102
+        if self._hooks:
+            logger.info(f"{len(self._hooks)} {self} hooks called with '{value}'")
+
         for hook in self._hooks:
             hook(self._hierarchy, value)
 
 
+class ProjectURL:
+    """Hook to add a project url to a build system."""
+
+    _hooks = []
+
+    @classmethod
+    def register(cls, hook):
+        """Regsiter a callback for :class:`ProjectURL` hook.
+
+        :param hook: Callaback to register.
+        :type hook: :class:`collections.abc.Callable`
+        """
+        cls._hooks.append(hook)
+
+    def __init__(self, hierarchy):
+        self._hierarchy = hierarchy
+
+    def __str__(self):
+        return "project-url"
+
+    def __call__(self, url_kind, url_value):  # noqa: D102
+        if self._hooks:
+            logger.info(
+                f"{len(self._hooks)} {self} hooks called with '{url_kind}: {url_value}'"
+            )
+
+        for hook in self._hooks:
+            hook(self._hierarchy, url_kind, url_value)
```

Comment:
To avoid repeating those idiom in each hook `register`, `__call__` and `__init__` method, maybe hooks should inherit from a `Hook` superclass, e.g:

Note `Hook` in not an ABC because among other things it provides concrete implementation of `__call__`.

```py
class Hook:
    _hooks = []

    @classmethod
    def register(cls, hook):
        cls._hooks.append(hook)

    def __init__(self, hierarchy):
        self._hierarchy = hierarchy

    def __str__(self):
        return self.__class__.__name__

    def __call__(self, *args, **kwargs):
        if self._hooks:
            logger.info(
                f"len(self._hooks)} {self} hooks called with parameters {args}, {kwargs}"
                # Maybe we can think of a nice way for child classes to redefine this logging message
            )
        
        for hook in self._hooks:
            hook(self._hierarchy, *args, **kwargs)


class ProjectURL(Hook):
    def __str__(self):
        return "project-url"
    
    def __call__(self, url_kind, url_value):
        """Here we can provide docs and relevant hook call signature"""
        
        super().__call__(url_kind, url_value)
```

Reply:
Good idea, can you try to implement something in that sense ? You can directly commit on the branch.